src/server/worker/master.go

```diff
@@ -473,6 +506,14 @@ func (a *APIServer) waitJob(pachClient *client.APIClient, jobInfo *pps.JobInfo,
 					if err := a.jobs.ReadWrite(stm).Get(jobInfo.Job.ID, jobPtr); err != nil {
 						return err
 					}
+					if jobInfo.EnableStats {
```

Comment:
Is there a reason why this cannot be done by StopJob? If the stats commit is set up at job creation time, then the race condition I was concerned about no longer exists.

Reply:
The lifetime of a job is defined by the lifetime of its output commit, so closing the output commit is the way to signal to kill the job, whereas the closing the stats commit is cleanup because the job has been killed. There are still some tricky conditions possible here if `StopJob` closes both commits, because it has to close one of the commits first, and thus it's possible, if it crashes at the right time, that one of them winds up closed while the other isn't. It's pretty unlikely but it's possible. So I think this way is slightly safer, and makes more conceptual sense, at least to me, because all that `StopJob` does is close the output commit, which is an atomic operation that signals the job has been killed, and then all of the cleanup is done by the master here.

Reply:
Closing the stats commit first guarantees that an output commit will not be closed without having stats closed. Closing the output commit first is why you need that recovery code for closing the stats commit when handling the terminal job state in the master. I think it is safer to close the stats commit first because the job will not be viewed as complete if it is still open (because the output commit will not be closed). I am fine either way because there are still race conditions with both approaches, and the approach you chose has a recovery method for crashing after closing the output commit but before closing stats. One race condition would end up having state from the job in one commit, but not the other (finish one commit while stopping, and the other in the regular job logic).